### PR TITLE
Jats oa and subtitles.20230714

### PR DESCRIPTION
**Pull request by seasidesparrow**
- More comprehensive parsing of licenses
- Better handling of multi-paragraph abstracts (e.g. A&A)
- Suppression of BeautifulSoup/JATS parsing warnings involving URI/URL markup

### CODE REVIEW

**Review by kelockhart — CHANGES_REQUESTED** [review]

File: adsingestp/parsers/jats.py [hunk 1]
```diff
@@ -11,6 +11,15 @@
...
 logger = logging.getLogger(__name__)
 
+# Ignore the specific warning that BeautifulSoup throws if it finds a URL
+# -- we will never need to pass any url to an HTTP client during parsing!
+try:
+    import warnings
+
+    warnings.filterwarnings("ignore", category=MarkupResemblesLocatorWarning, module="bs4")
+except Exception as err:
+    logger.warning("Problem filtering BeautifulSoup warnings: %s" % err)
```
[the remark this answers is not in the document]
Comment (seasidesparrow) [reply adsingestp/parsers/jats.py:21]: In the bs4 codebase: https://git.launchpad.net/beautifulsoup/tree/bs4/__init__.py#n411

License indicators are typically given as URLs to the website of the license (e.g. https://www.creativecommons.org/...license...), so when it parses that, it throws the warning linked above.  It's not a fatal error, just a (confusing) warning.

File: adsingestp/parsers/crossref.py [hunk 2]
```diff
@@ -221,6 +221,11 @@ def _parse_title_abstract(self):
         if self.record_meta.find("titles") and self.record_meta.find("titles").find("title"):
             self.base_metadata["title"] = self.record_meta.find("titles").find("title").get_text()
 
+        if self.record_meta.find("titles") and self.record_meta.find("titles").find("subtitle"):
+            self.base_metadata["subtitle"] = (
+                self.record_meta.find("titles").find("subtitle").get_text()
+            )
```
[the remark this answers is not in the document]
Comment (seasidesparrow) [reply adsingestp/parsers/crossref.py:227]: No.  I'll fix.

File: tests/stubdata/input/jats_a+a_subtitle.xml [hunk 3]
```diff
@@ -0,0 +1,5094 @@
...
+<meta-name>open-access</meta-name>
+<meta-value>yes</meta-value>
+</custom-meta>
+<custom-meta><meta-name>transformative_agreement</meta-name><meta-value>national-agreement-fr_2022</meta-value></custom-meta></custom-meta-group>
+</article-meta>
+</front>
+<body>
+<sec id="S1" sec-type="intro">
+<label>1.</label>
+<title>Introduction</title>
```
[the remark this answers is not in the document]
Comment (seasidesparrow) [reply tests/stubdata/input/jats_a+a_subtitle.xml:606]: Yes, it's CC4.0.

File: tests/stubdata/output/jats_aip_aipc_2470_040010.json [hunk 4]
```diff
@@ -1,34 +1,188 @@
...
-                "<ref id=\"c17\"><label>17.</label><mixed-citation publication-type=\"journal\"><string-name><given-names>A.M.</given-names> <surname>Farías-Campomanes</surname></string-name>, <string-name><given-names>C. N.</given-names> <surname>Horita</surname></string-name>, <string-name><given-names>M.A.R.</given-names> <surname>Pollonio</surname></string-name>, <string-name><given-names>M. A. A.</given-names> <surname>Meireles</surname></string-name>. <article-title>Allicin-Rich Extract Obtained from Garlic by Pressurized Liquid Extraction: Quantitative Determination of Allicin in Garlic Samples</article-title>, <source>Food and Public Health</source> <volume>4</volume>(<issue>6</issue>): <fpage>272</fpage>–<lpage>278</lpage> (<year>2014</year>)</mixed-citation></ref>"],
- "title": {"textEnglish": "Synthesis and characterization of chitosan-Allium sativum film"}}
+{
```
[the remark this answers is not in the document]
Comment (seasidesparrow) [reply tests/stubdata/output/jats_aip_aipc_2470_040010.json:1]: I'm using `json.dumps(data, indent=2)`.

File: tests/stubdata/output/crossref_cn_10.1051=0004-6361=202243540.json [hunk 5]
```diff
@@ -1 +1 @@
...
...Tsuji</author> <volume>23</volume> <first_page>411</first_page> <cYear>1973</cYear> </citation>", "<citation key=\"R97\"> <journal_title>A&A</journal_title> <author>van der Helm</author> <volume>625</volume> <first_page>A85</first_page> <cYear>2019</cYear> <doi provider=\"crossref\">10.1051/0004-6361/201732020</doi> </citation>", "<citation key=\"R98\"> <journal_title>Ap&SS</journal_title> <author>Velli</author> <volume>277</volume> <first_page>157</first_page> <cYear>2001</cYear> <doi provider=\"crossref\">10.1023/A:1012237708634</doi> </citation>", "<citation key=\"R99\"> <journal_title>SIAM J. Sci. Comput.</journal_title> <author>Wang</author> <volume>33</volume> <first_page>2536</first_page> <cYear>2011</cYear> <doi provider=\"crossref\">10.1137/090761355</doi> </citation>", "<citation key=\"R100\"> <journal_title>MNRAS</journal_title> <author>Whitehouse</author> <volume>364</volume> <first_page>1367</first_page> <cYear>2005</cYear> <doi provider=\"crossref\">10.1111/j.1365-2966.2005.09683.x</doi> </citation>", "<citation key=\"R101\"> <journal_title>A&A</journal_title> <author>Winters</author> <volume>361</volume> <first_page>641</first_page> <cYear>2000</cYear> </citation>"]}
```
[the remark this answers is not in the document]
Comment (seasidesparrow) [reply tests/stubdata/output/crossref_cn_10.1051=0004-6361=202243540.json:1]: I reformatted this and two other files, using `json.dumps(data, indent=2)`.

File: adsingestp/parsers/jats.py [hunk 6]
```diff
@@ -526,21 +537,44 @@ def _get_date(self, d):
...
+                    subtitle = title_group.find("subtitle")
+                    for dx in subtitle.find_all("ext-link"):
+                        self.subtitledoi = dx.find("xlink:href")
+                    for dx in subtitle.find_all("xref"):
+                        subtitle_xref_list.append(
+                            self._detag(dx, self.JATS_TAGSET["abstract"]).strip()
+                        )
+                        dx.decompose()
+                    for df in subtitle.find_all("fn"):
+                        subtitle_fn_list.append(
```
[the remark this answers is not in the document]
Comment (seasidesparrow) [reply adsingestp/parsers/jats.py:568]: I don't have example of a document that has this, but the JATS 1.3 reference says that `<fn>` may be contained in `<subtitle>` (https://jats.nlm.nih.gov/archiving/tag-library/1.3/element/fn.html).  I added a second if statement immediately after L590 to do the same thing with `subtitle_fn_list`

File: adsingestp/parsers/jats.py [hunk 7]
```diff
@@ -555,6 +555,7 @@ def _parse_title_abstract(self):
                 for df in title.find_all("fn"):
                     title_fn_list.append(self._detag(df, self.JATS_TAGSET["abstract"]).strip())
                     df.decompose()
+                art_title = self._detag(title, self.JATS_TAGSET["title"]).strip()
                 if title_group.find("subtitle"):
                     subtitle = title_group.find("subtitle")
                     for dx in subtitle.find_all("ext-link"):
```
Comment (kelockhart) [inline adsingestp/parsers/jats.py:561]: Are `self.subtitledoi` and `subtitle_xref_list` used anywhere? If not, this logic can be removed

**Review by kelockhart — CHANGES_REQUESTED** [review]
Just a couple small comments on the subtitles that got missed in the first round, but looks good otherwise.

File: adsingestp/parsers/jats.py (hunk 7 above)
Comment (seasidesparrow) [reply adsingestp/parsers/jats.py:561]: subtitledoi can be removed.  Unfortunately there can be xrefs with subtitles, as in `stubdata/input/jats_a+a_subtitle.xml` but there's no room in the subtitle data element in adsingestschema.Document, it's just plain text.  In classic, title footnotes are appended to the abstract (not title), and I'm going to confirm with CSG if that's still behavior we want to continue.  I'd rather update ingestschema to add a structure for subtitle data; if we want it to be merged into the abstract I think that should happen downstream during enrichment, not parsing.

Comment (seasidesparrow) [reply adsingestp/parsers/jats.py]: This has now been resolved and tested.

**Comment by seasidesparrow** [comment]
Latest commits include adding funding capture to jats parser

**Review by kelockhart — APPROVED** [review]
Looks good! Merge when ready.

**Pull request merged**